File: src/chapters/iterative-methods/content/code/gauss-seidel.go
package main

import (
	"fmt"
	"math"
)

// Solve A x = b by Gauss-Seidel iteration. Returns the solution and iteration count.
func gaussSeidel(A [][]float64, b []float64, tol float64, maxIter int) ([]float64, int) {
	n := len(b)
	x := make([]float64, n)
	for k := 1; k <= maxIter; k++ {
		diff := 0.0
		for i := 0; i < n; i++ {
			s := b[i]
			for j := 0; j < n; j++ {
				if j != i {
					s -= A[i][j] * x[j]
				}
			}
			xi := s / A[i][i]
			diff = math.Max(diff, math.Abs(xi-x[i]))
			x[i] = xi
		}
		if diff <= tol {
			return x, k
		}
	}
	return x, maxIter
}

func main() {
	A := [][]float64{{4, 2, -1}, {5, -10, 2}, {-2, 3, -7}}
	b := []float64{9, 8, 3}
	x, it := gaussSeidel(A, b, 1e-10, 200)
	fmt.Println("x =", x, " iterations =", it)
}

File: src/chapters/iterative-methods/content/code/gauss-seidel.js
// Solve A x = b by Gauss-Seidel iteration. Returns [x, iterations].
function gaussSeidel(A, b, tol = 1e-10, maxIter = 200) {
  const n = b.length;
  const x = new Array(n).fill(0);
  for (let k = 1; k <= maxIter; k++) {
    let diff = 0;
    for (let i = 0; i < n; i++) {
      let s = b[i];
      for (let j = 0; j < n; j++) if (j !== i) s -= A[i][j] * x[j];
      const xi = s / A[i][i];
      diff = Math.max(diff, Math.abs(xi - x[i]));
      x[i] = xi;
    }
    if (diff <= tol) return [x, k];
  }
  return [x, maxIter];
}

const A = [[4, 2, -1], [5, -10, 2], [-2, 3, -7]];
const b = [9, 8, 3];
const [x, it] = gaussSeidel(A, b);
console.log("x =", x, " iterations =", it);
